### DISCUSSION TITLE
Public Inputs for SNARK Accounts

### DISCUSSION THREAD

**Discussion by evan-forbes** [celestiaorg/celestia-app#2902] [Ideas]
Typical zk-rollups utilize smart contracts to enforce what must be provided as a public input when verifying a proof. Since smart contracts are not an option, we must define sufficiently general required public inputs.

In the forum post, it suggests that

> the public input to the rollup’s Celestia ZK address would be the last Celestia block height (which would be enforced as an input by Celestia’s state machine), and the identifier of some “withdrawal transaction” on the rollup.

However AFAIU, this would require the Celestia state machine to parse the withdrawal transaction to check that the transaction controlling the SNARK Account is actually sending the correct amount of funds to the correct address.

Instead, I think we could actually pass the hash of the transaction (or the hash of the body of the transaction) as a public input. This would allow for much more arbitrary actions to be proven. The circuit would first prove that some tx bytes hash to some value, and then prove arbitrary things about those bytes. Such as they contain only the correct bank.MsgSends.

There might be values that would be very useful to include as [optional] public inputs. Recall that a rollup would have to prove the fork choice rule and inclusion of some data. This could be done by first proving a light celestia light client, and then using that light client to verify data roots. Those data roots could then be used to verify inclusion and or the fork choice rule.

Alternatively, the state machine could require that some commitment to block data be included as a public input. This commitment could be the raw data roots themselves, or perhaps some meta commitment such as what is used for blobstream.

**Comment by musalbas** [comment]
I only meant that the block height is the only consensus-enforced public input; the rest can be arbitrary.

**Comment by staheri14** [comment]
Based on the discussions we've had so far in https://github.com/celestiaorg/celestia-app/discussions/2904 , it appears infeasible to establish a universal predicate applicable to all rollups. Consequently, while it's not possible to determine the exact public inputs required by each rollup, it is essential to anticipate potentially useful public inputs and provide them to the rollups predicate verification function. _Rollups will have the option to utilize or disregard these inputs_. _However, they are expected to accept these inputs in their validity predicate/circuit._ 

Such public inputs include:

- `tx_hash`: This is the hash of the withdrawal transaction intended for the SNARK account. The specific fields of the transaction that should be included in the hash computation need to be specified. However, the amount(s) and the address(es) of the receiver account(s) are definite inclusions. The aim is for Celestia to associate each transaction with its corresponding proof, thereby preventing replay attacks.
- The data root of the `i`th block of Celestia, where `i` is determined by the withdrawal transaction. (a general case covering what mentioned in  this [comment](https://github.com/celestiaorg/celestia-app/discussions/2902#discussioncomment-7781171) as well)
- The block `d'` (where `d'` can be the blob share commitment corresponding to the rollups data submitted to Celestia). `d'` is expected to be included in the i-th block of Celestia (as explained in [here](https://celestiaorg.github.io/celestia-app/specs/data_square_layout.html?highlight=share%20commitment#blob-share-commitment-rules)).

**Reply by staheri14** [reply]
Another public input that is needed is a _nullifier_, to prevent double spending from the SNARK account. 
Since this account is holding funds related to a rollup, it needs to protected against double spending/withdrawal where one single valid proof can be used to transfer funds multiple times. This can be achieved through a nullifier. This approach could effectively eliminate the risk of double spending. One such nullifier could be roughly `H(d',receiver_account)`, given that at each block `d'` of the rollup there will only be one valid withdrawal from the SNARK account to the `receiver_account`.

**Reply by musalbas** [reply]
You're right that we need to prevent double spending within the same block, which could be achieved by a consensus-critical rule that allows only one valid withdrawal per block corresponding to a specific withdrawal transaction identifier, or receiver account.

We need to differentiate between consensus-critical inputs and non-consensus-critical inputs though. I think the only consensus-critical inputs are the block height and the receiver account (to prevent double spends). The data root and block data could be parsed by the ZK rollup itself. We should also discuss if we want to make any inputs consensus-critical that don't need to be, in the interests of making the ZK rollups easier to construct, if they don't add too much overhead to the Celestia state machine.

**Reply by staheri14** [reply]
>  We should also discuss if we want to make any inputs consensus-critical that don't need to be, in the interests of making the ZK rollups easier to construct, if they don't add too much overhead to the Celestia state machine.

Regarding the previously mentioned aspect (above), persisting valid rollup state transitions within the SNARK account state would be beneficial. Here's the explanation.

When a SNARK transaction is received, it claims that due to the latest block `B'i` of the rollup, a state `S'i` has resulted, necessitating a transfer of some `amount` to a `receiver account`. In the snark validity predicate, at least two things must be verified:
- `B’i` is the canonical block of the rollup.
- `S’i `is the correct state of the rollup, considering its entire history up to `B’i`.
For accurate verification of these points, it is likely necessary for the entire history of the rollup up to block `i` to be submitted by the transaction owner and verified in the predicate. This is because the snark account does not hold the rollup's recent state (and has no notion of what is the most recent valid state of the rollup), essential for verifying the state transition against the most recent one.

This approach could lead to potenital issues: the number of public inputs will grow significantly, resulting in larger transaction sizes as these inputs are part of the transaction fields. Additionally, the validity predicate would become too large increasing proof time.

One way to address this issue is to allow valid state transitions to be persisted in the snark account state. Therefore, the snark validity predicate would only need to verify the transition from the latest known state to the next one, instead of the entire rollup history.

Determining which state transition is valid for persistence can be approached in two ways:

1. This information can be extracted from the incoming withdrawal transaction. For example, if an incoming snark transaction claims that an `amount` should be withdrawn from the snark account to a `receiver_account` due to the presence of a block `B’i` in the rollup that transitioned the state from `S’i-1` to `S’i`. If the proof attached to the transaction is valid (can be verified on the celestia side), it indicates `S'i` is a valid state transition and can be persisted in the snark account state.
2. In a different method, the snark account could be associated with an additional predicate representing the state transition validity rules. State updates can occur by submitting a different transaction that carries the latest state hash `S'i-1`, the recent block hash `B'i`, and the new state hash `S'i`, along with a different `proof`. If the `proof` is verified, then the new state hash `S_i` will be persisted as the canonical one.

For other incoming withdrawal transactions that point to the a block height `i` of the rollup, i.e.,` S'i` (which is already available in the account state), the snark account’s state is retrieved and compared with the transaction’s `S'i`. If they do not match, the transaction is considered invalid.

**Reply by musalbas** [reply]
>For accurate verification of these points, it is likely necessary for the entire history of the rollup up to block i to be submitted by the transaction owner and verified in the predicate.

As [discussed here](https://github.com/celestiaorg/celestia-app/discussions/2904#discussioncomment-7852606), can't you just use recursive SNARK proofs that build on the previous proofs? The previous proofs could perhaps have some Merkle root of pending withdrawals as an input. We could perhaps have the post-withdrawal root be updated implicitly in the next block; rather than immediately after the withdrawal transaction.

The issue with adding post-state roots for each withdrawal is that only one withdrawal transaction can processed at a time, as I assume that each withdrawal would need a lock on the entire state.

**Reply by staheri14** [reply]
Could you clarify why state locking is necessary? It seems to me that locking the state isn't required. This is my pov: each rollup state S'i specifies certain balance adjustments for specific Celestia accounts. For example, x amount is allocated to Alice's account, and y amount to Bob's. Each rollup state reflects the conclusive balance after processing all transactions within that specific rollup block. Therefore, it represents a unique update for each involved account (like Alice's and Bob's here). Given this, locking the state S'i seems unnecessary, allowing for the possibility of concurrent withdrawals.

Update: I think you're suggesting that when a rollup state update occurs, not all withdrawals resulting from that update happen immediately on celestia. Therefore, withdrawals resulting from subsequent rollup states must wait until all previous rollup states are fully executed. Right?
However, if the state changes on the rollup are consistent and non-conflicting, then state locking again does not look necessary.

**Reply by musalbas** [reply]
This issue is that if each withdrawal updates the state root of the rollup, then don't you need a lock on the state root, thus the entire state of the rollup?

**Reply by staheri14** [reply]
If states were to be updated per withdrawal tx then state locking is needed, however, I think the following approach would eliminate the need for state locking, please let me know your POV. 

- For every rollup block, `Bi`, the state transitions from `S_i-1` to `Si` and is recorded. `Si` represents the state of the rollup after executing **all** transactions in `Bi`.   For every rollup state transition, some celestia accounts are entitled to receive some funds. For example, a Celestia account acc1 might be designated to receive 3 TIA, and another account acc5 might be set to receive 10 TIA. The withdrawal transactions that are eligible due to this state transition are **not yet submitted to celestia and will arrive later**.

- Once the state transition is recorded, withdrawal transactions can be submitted. These can be processed either in batches or individually. The key is that these transactions will reference the state transition, proving that based on this transition, e.g., acc1 is entitled to withdraw 3 TIA, and acc5 can withdraw 10 TIA.  

By following this approach, state locking doesn't seem necessary because the state transitions are recorded and acknowledged before the actual withdrawal transactions are submitted. Also, withdrawal transactions referencing different state transitions can  arrive out of order as they point to their own particular state transition which is already proven to be correct.

**Reply by musalbas** [reply]
This is possible but this would be a different (and more state-heavy) design than the one originally proposed, where each rollup has an Ethereum-style contract with an "on-chain light client" where its state is updated. The tradeoffs between these designs should be evaluated, i.e. this could be more state-heavy but perhaps require less computation by the ZK prover for people building non-based rollups

**Reply by puma314** [reply]
1 possible problem I see with the rollup `tx_hash` being used to prevent replay attacks is that on general-purpose execution environments, 1 transaction on the rollup can have multiple "withdraws" within that transaction (for example imagine that you make a smart contract that calls the "withdraw" contract on the rollup 10 times with 10 different withdrawal addresses). I think it might be hard to prevent this behavior on rollups (or something that is bad to assume in general), so I think just using the transaction hash is not sufficient (at least on the rollup -> Celestia direction).

On Celestia, my understanding is that we can make the assumption that 1 transaction = 1 rollup deposit due to the limited nature of transaction types, right?

**Reply by puma314** [reply]
Also @staheri14 is the idea with your proposed design that the Snark Account's state root gets updated every time a blob of transactions pertaining to the related rollup gets posted on Celestia? Otherwise, in the SNARK account don't you have to keep track of the last processed Celestia block height and then show that if you include all blocks in between the last_processed height and the current height and filter for blobs in a particular namespace that it results in a certain batch of transactions to apply the rollup STF on? 

This seems pretty computationally expensive to do in a proof since you have to iterate through all transactions in Celestia blocks potentially.

**Reply by musalbas** [reply]
You should be able to iterate through all the data in a namespace (by traversing the Namespace Merkle Tree) without filtering all the transactions. However, you may need to filter all the transactions to parse relevant Celestia deposit or withdrawal transactions, unless we change the way they are structured or ordered in the block.

**Reply by puma314** [reply]
When traversing the Namespace Merkle Tree, can you prove that a blob is the `i`-th entry in the tree (assuming it is the `i`-th blob posted to that namespace)? And is that proof just a standard merkle proof, so proving items `i` through `j` will take `log_n * (j-i)` time?

**Reply by musalbas** [reply]
You can't prove that a blob is the `i`-th entry but you can prove that a share is the `i`-th share. Why do you to prove a blob is the `i`-th entry?

**Reply by bakugo49** [reply]
Without proving that the blob is the `i`-th entry, how does the SNARK account verify that no blobs have been skipped over by the prover in the proof?

**Reply by musalbas** [reply]
if you know the namespace starts at share index x for example, and ends at index x+y, then you check no shares were skipped over between  x and x+y

**Reply by evan-forbes** [reply]
> 1 transaction on the rollup can have multiple "withdraws" within that transaction (for example imagine that you make a smart contract that calls the "withdraw" contract on the rollup 10 times with 10 different withdrawal addresses). I think it might be hard to prevent this behavior on rollups (or something that is bad to assume in general), so I think just using the transaction hash is not sufficient (at least on the rollup -> Celestia direction).

@puma314 do you mind expanding on this point? I'm not yet following why the tx hash could be used for a single withdraw, but not multiple 


> On Celestia, my understanding is that we can make the assumption that 1 transaction = 1 rollup deposit due to the limited nature of transaction types, right?

unfortunately, we can't make this assumption. we have have a module that enables authorization of other accounts to perform certain actions. Since a transaction can contain an arbitrary number of actions, its possible or a single authorized account to send funds for an arbitrary number of accounts in a single transaction.

**Comment by staheri14** [comment]
I’d like to present a classification of public inputs, which I've found beneficial in designing and managing the workflow of snark accounts and the current and future discussions:

| Classification                     | Description                                                                                                                                                      | Example                                    | Belongs to | 
|------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------|--------------------------------------------|
| **Static Inputs**                  | These inputs do not change over time; are provided by the account holder at the time of account creation.                                                                                           | Maximum withdrawal amount for each transaction (if applicable)| Account state |
| **Dynamic Inputs**                 | Inputs that are time dependant. They are supplied at the time of submitting withdrawal transactions.                                                                                               | Data root of the ith Celestia block          | Transaction fields |

The **Dynamic Inputs** can be further grouped: 
| Classification  of Dynamic Inputs                | Description                                                                                                                                                      | Example                                    | 
|------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------|
| **Mutual Inputs**    | Such public inputs are mutually understandable by the rollup users (account holder/tx owners) and celestia.  They may be supplied by celestia or the rollup (i.e., tx owner). |           tx_hash, d, d' |                        
| **Rollup-Specific Inputs** | Such inputs are rollup specific; the interpretation of these inputs are outside the scope of SNARK account design.                            |  anything |

And the last classification is about the **Mutual Inputs**:
| Classification of Mutual Inputs                   | Belongs to   |
|------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------| 
| **Supplied by Rollup/Transaction Owner**    | Part of the transaction fields.                                                                                                                                    | 
| **Supplied by Celestia**             | Omitted from the transaction fields.                                                                                                                           | 

With the presented classification, I'd like to draw several observations and design ideas:

1. The public inputs for the SNARK verification algorithm fall into these four categories:

- Static Inputs
- Dynamic Mutual Inputs Supplied by Celestia
- Dynamic Mutual Inputs Supplied by Rollup
- Dynamic Rollup-Specific Inputs

2. For verifying a SNARK proof, these four classes of public inputs should be serialized and positioned correctly before being fed into the verification algorithm (`vk` stands for verification key):
`Verify(vk, proof, [Dynamic Mutual Inputs Supplied by Celestia] || [Dynamic Mutual Inputs Supplied by Rollup] || [Static Inputs] || [Dynamic Rollup-Specific Inputs])`.  

3. Given that the number of public inputs varies based on the SNARK predicate, the account holder must specify the total number of public inputs at the time of account creation. This value is static and should be persisted as part of the account's state.
4. The mutual inputs provided by Celestia could be predefined. Therefore, all SNARK accounts' circuits should be designed to accept these inputs at a specific position in their circuit. It is up to the SNARK account holder to either discard or utilize these inputs in the SNARK validity predicate.
5. Similarly, there could be a predefined list of mutual inputs provided by the rollup or transaction owner. These inputs will form additional SNARK-related and named fields of a withdrawal transaction. Celestia may perform correctness checks on such inputs for additional security and integrity.

**Reply by staheri14** [reply]
cc: @cmwaters follow-up on https://github.com/celestiaorg/celestia-app/discussions/2904#discussioncomment-7846814
